### DISCUSSION TITLE
Frontend for another language for JSImpl

### DISCUSSION THREAD

**Discussion by dtrendafilov** [SofiaCPP/IPL#105] [Projects]
Create a lexer and a parser for another language that will produce the AST in the same JSON format as JSImpl, so that Spasm can be used to interpret that language.

- Using an existing parser and translating its output to the JSImpl JSON will be considered too easy.

You can also skip the JSON step by directly using the AST C++ classes.

So there are two approaches:

1. Produce JSON. The format 
 is basically serialization of the AST. You can see the [JSON Printer](https://github.com/SofiaCPP/IPL/blob/ba9f446bcb4a5ec55f29814034b66df4d051fae9/JSImpl/src/ASTPrinter.cpp#L105)
 
 2. Fork the repo, add a folder `<Language>` with a lexer and parser inside and directly use the  [AST in JSImpl](https://github.com/SofiaCPP/IPL/blob/ba9f446bcb4a5ec55f29814034b66df4d051fae9/JSImpl/src/ExpressionDefinitions.h#L90)

**Comment by MariyanMomchilov** [comment]
Здравейте, 
 
Искам да направя frontend за Lua, ако е подходящо за проект :)

**Reply by NDobrev** [reply]
Ок е

**Comment by yzelova** [comment]
Здравейте,

Аз искам да направя frontend за Rust.

**Reply by dtrendafilov** [reply]
Rust е статично типизиран, което JSImpl AST-то не поддържа out of the box. Вариант е да игнорираш типовете и да стане динамично типизиран Rust.

**Reply by yzelova** [reply]
Ако е подходящо за проект, така ще го направя

**Reply by dtrendafilov** [reply]
Става, да.

**Comment by HristianIliev** [comment]
Аз бих искал да правя този проект. Ще правя фронтенд за Python.

Междудругото може ли да постнете example output за това как трябва да изглежда AST-то задължително.

**Reply by dtrendafilov** [reply]
Добавих линкове в описанието за JSON формата и възможността да се използва директно C++ класовете.

**Comment by dimitar10000** [comment]
Аз също искам да правя този проект за C++, но ще трябва да знам как ще изглежда AST? Дайте повече информация, ако може.

**Reply by dtrendafilov** [reply]
Добавих линкове в описанието за JSON формата и възможността да се използва директно C++ класовете.

Целия C++ или някаква част от C++?

**Reply by dimitar10000** [reply]
Ами не целия, ще се опитам колкото може повече неща да направя, но няма да има всичко.

**Comment by Stuksi** [comment]
Здравейте, бих искал да направя frontend за Lisp.

**Reply by dtrendafilov** [reply]
В Lisp има ли достатъчно синтактични конструкции? Не е ли изрази от скоби само?

**Reply by Stuksi** [reply]
Така е, доста е символичен. ~~За Lua в такъв случай?~~ За Lua е имало вече, за Ruby?

**Reply by dtrendafilov** [reply]
Ruby 💎 е ОК.

**Comment by ConDev20** [comment]
Аз също искам да правя този проект. Към момента не съм се спрял на конкретен език. Бих се радвал на предложения.

**Reply by dtrendafilov** [reply]
За предпочитане някой език със синтаксис поне колкото C, ако не и повече.

**Reply by ConDev20** [reply]
Php окей ли е ?

**Reply by dtrendafilov** [reply]
:+1:
